code/game/machinery/station_map.dm

```diff
@@ -0,0 +1,321 @@
+
+/obj/machinery/station_map/power_change()
+	. = ..()
+	update_icon()
+
+/obj/machinery/station_map/proc/set_broken()
+	stat |= BROKEN
+	update_icon()
+
+/obj/machinery/station_map/update_icon()
+	overlays.len = 0
+	if(stat & BROKEN)
+		icon_state = "station_mapb"
+	else if((stat & NOPOWER) || !anchored)
+		icon_state = "station_map0"
+	else
+		icon_state = "station_map"
+
+		if(bogus)
+			station_map.overlays.len = 0
+			station_map.overlays |= legend
+		else
+			switch(dir)
+				if(NORTH)
+					small_station_map.icon = extraMiniMaps[HOLOMAP_EXTRA_STATIONMAPSMALL_NORTH+"_[original_zLevel]"]
+				if(SOUTH)
+					small_station_map.icon = extraMiniMaps[HOLOMAP_EXTRA_STATIONMAPSMALL_SOUTH+"_[original_zLevel]"]
+				if(EAST)
+					small_station_map.icon = extraMiniMaps[HOLOMAP_EXTRA_STATIONMAPSMALL_EAST+"_[original_zLevel]"]
+				if(WEST)
+					small_station_map.icon = extraMiniMaps[HOLOMAP_EXTRA_STATIONMAPSMALL_WEST+"_[original_zLevel]"]
+
+			overlays |= small_station_map
+			station_map.overlays.len = 0
+			if(map.holomap_offset_x.len >= original_zLevel)
+				cursor.pixel_x = (x-6+map.holomap_offset_x[original_zLevel])*PIXEL_MULTIPLIER
+				cursor.pixel_y = (y-6+map.holomap_offset_y[original_zLevel])*PIXEL_MULTIPLIER
+			else
+				cursor.pixel_x = (x-6)*PIXEL_MULTIPLIER
+				cursor.pixel_y = (y-6)*PIXEL_MULTIPLIER
+			station_map.overlays |= cursor
+			station_map.overlays |= legend
+
+	switch(dir)
+		if(NORTH)
+			pixel_x = 0
+			pixel_y = WORLD_ICON_SIZE
+		if(SOUTH)
+			pixel_x = 0
+			pixel_y = -1*WORLD_ICON_SIZE
+		if(EAST)
+			pixel_x = WORLD_ICON_SIZE
+			pixel_y = 0
+		if(WEST)
+			pixel_x = -1*WORLD_ICON_SIZE
+			pixel_y = 0
+
+	if(floor_markings)
+		floor_markings.dir = dir
+		floor_markings.pixel_x = -1*pixel_x
+		floor_markings.pixel_y = -1*pixel_y
+		overlays |= floor_markings
+
+	if(panel_open)
+		overlays |= "station_map-panel"
+	else
+		overlays -= "station_map-panel"
+
+/obj/machinery/station_map/ex_act(severity)
+	switch(severity)
+		if(1.0)
+			qdel(src)
+		if(2.0)
+			if (prob(50))
+				qdel(src)
+			else
+				set_broken()
+		if(3.0)
+			if (prob(25))
+				set_broken()
+
+//Portable holomaps, currently AI/Borg/MoMMI only
+/obj/item/device/station_map
+	name					= "portable station holomap"
+	desc					= "A virtual map of the surrounding station."
+	icon_state				= "station_map"
+	flags					= FPRINT
+	siemens_coefficient		= 1
+	force					= 5.0
+	w_class					= 2.0
+	throwforce				= 5.0
+	throw_range				= 15
+	throw_speed				= 3
+	starting_materials		= list(MAT_IRON = 50, MAT_GLASS = 20)
+	w_type					= RECYK_ELECTRONIC
+	melt_temperature		= MELTPOINT_SILICON
+	origin_tech				= Tc_MAGNETS + "=2;" + Tc_PROGRAMMING + "=2"
+
+	var/mob/watching_mob = null
+	var/image/station_map = null
+	var/image/cursor = null
+	var/image/legend = null
+
+/obj/item/device/station_map/attack_self(var/mob/user)
+	toggleHolomap(user)
+
+/obj/item/device/station_map/proc/toggleHolomap(var/mob/user,var/isAI=0)
```

Comment:
Make a specialized datum that has the code for holomaps like this instead, then use that to remove copy paste.